### PR TITLE
Introduce deploy CLI commands

### PR DESCRIPTION
**Pull request by iocanel**
This pull request exposes the deployment capabilities of Quarkus across all our dev tools (maven, gradle, cli).

## Deployment from the CLI
Introduces the `deploy` command  as well as specific subcommands for:
- kubernetes
  - kind
  - minikube
- knative
- openshift

Simililarly to the `image` commands, it allows users to perform actions without requiring them to modify their pom to bring in the needed extension.

### CODE REVIEW

**Review by geoand — COMMENTED** [review]
Very interesting!

I can't comment on the CLI stuff, but the k8s stuff looks nice :)

**Comment by maxandersen** [comment]
really interesting - i'll look; would be great if @ebullient or @aloubyansky could chime in if anything alarming :)

File: core/deployment/src/main/java/io/quarkus/deployment/util/DeploymentUtil.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+     * The list is obtained by checking for properties `quarkus.xxx.deploy`.
+     * These properties have a default value and thus they will be found regardless of the
+     * actual user configuration, so we check for property names instead.
+     *
+     * @return a {@link List} with all available deployers.
+     */
+    public static List<String> getDeployers() {
+        Config config = ConfigProvider.getConfig();
+        return Stream.concat(System.getProperties().entrySet().stream().map(e -> String.valueOf(e.getKey())),
+                StreamSupport.stream(config.getPropertyNames().spliterator(),
```
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/util/DeploymentUtil.java:32]: Shouldn't `config` already include `System.getProperties()`?

Comment (iocanel) [reply core/deployment/src/main/java/io/quarkus/deployment/util/DeploymentUtil.java:32]: I think that this was not that case, but if you are positive I can have a second look.

**Review by aloubyansky — APPROVED** [review]
Added a minor comment but otherwise looks good to me. Great work @iocanel!

File: devtools/cli/src/main/java/io/quarkus/cli/deploy/KubernetesOptions.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
+package io.quarkus.cli.deploy;
+
+import java.util.Optional;
+
+import picocli.CommandLine;
+
+public class KubernetesOptions {
+
+    @CommandLine.Option(order = 3, names = { "--master-url" }, description = "The URL of the kubernetes API server")
+    public Optional<String> masterUrl;
```
[the remark this answers is not in the document]
Comment (iocanel) [reply devtools/cli/src/main/java/io/quarkus/cli/deploy/KubernetesOptions.java:10]: `master-url` is what we use for configuring the kubernetes-client in the kubernetes-client extension and also in the fabric8 kubernetes-client (see [1]). While, I agree that something like `apiserver-url` or `api-server-url` would be a better choice, I think it's better we stay consistent. 

[1]: https://github.com/quarkusio/quarkus/blob/290bf2428f3f695ef004668507407d30118e12fd/extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientBuildConfig.java#L24

[the remark this answers is not in the document]
Comment (ebullient) [reply devtools/cli/src/main/java/io/quarkus/cli/deploy/KubernetesOptions.java:10]: The push is to move away from this kind of language, and the CLI is hiding some other things. So, in the name of consistency, we should start deprecating/aliasing/moving the other parameter names... not persisting usage into the CLI, too.

**Review by ebullient — CHANGES_REQUESTED** [review]
`--apiserver-url` .. let's please move into more inclusive langauge. We can map the terms in help, and begin the transition in other tools, too.

File: core/deployment/src/main/java/io/quarkus/deployment/util/DeploymentUtil.java (hunk 1 above)
Comment (iocanel) [reply core/deployment/src/main/java/io/quarkus/deployment/util/DeploymentUtil.java]: Removed the System property check and retried all the scenarios. It seems to be working even without it. Updated the PR accordingly.

File: devtools/cli/src/main/java/io/quarkus/cli/deploy/KubernetesOptions.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (iocanel) [reply devtools/cli/src/main/java/io/quarkus/cli/deploy/KubernetesOptions.java:10]: Added a commit that deprecates the use of `master-url` in favor of `api-server-url`.

**Review by ebullient — APPROVED** [review]

**Comment by iocanel** [comment]
@gsmet, @geoand: the failure seem unrelated what should I do with this pr?

**Comment by geoand** [comment]
The failure indeed seems unrelated.

+1 for merging

**Comment by gsmet** [comment]
@iocanel I added the noteworthy-feature label. If you add a comment here with a blurb explaining the feature, I will happily add it to the announcement.

**Comment by gsmet** [comment]
Yes, thanks.

**Pull request merged**